### PR TITLE
Enable support for control plane metrics collection and publishing

### PR DESCRIPTION
**Pull request by ironcladlou**
This PR adds some foundational support to enable control plane metrics and
external publishing by:

1. Ensuring all control plane component pods are labeled with a new
`hypershift.openshift.io/control-plane-component` label which disambiguates them
from other pods in the cluster (which the `app` label is insufficient for).

2. Installing a new `PrometheusRule` in the `hypershift` namespace which defines a
recording rule to capture memory metrics for control plane pods and expose them
as a `hypershift:controlplane:component_memory_usage` metric.

3. Labeling the `hypershift` namespace appropriately so that the recording rule is
processed by the cluster monitoring stack if the stack is properly configured.

4. Adding a new `test/setup` command which provides tools usable by developers or
CI to configure a management cluster's monitoring stack (both cluster and user
workload) to capture all our metrics and optionally enable publishing them to a
Prometheus remote write endpoint.

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/pki/cert.go [hunk 1]
```diff
@@ -34,12 +34,14 @@ func reconcileSignedCertWithAddresses(secret, ca *corev1.Secret, ownerRef config
 }
 
 func reconcileSignedCertWithKeysAndAddresses(secret *corev1.Secret, ca *corev1.Secret, ownerRef config.OwnerRef, cn, org string, usage x509.KeyUsage, extUsages []x509.ExtKeyUsage, crtKey, keyKey, caKey string, dnsNames []string, ips []string) error {
+	if secret.CreationTimestamp.IsZero() {
+		secret.Type = corev1.SecretTypeOpaque
+	}
```
Comment (ironcladlou) [inline control-plane-operator/controllers/hostedcontrolplane/pki/cert.go:39]: @enxebre I needed this to fix crashing, probably need to split into another PR

File: Makefile [hunk 2]
```diff
@@ -15,6 +15,8 @@ GO_GCFLAGS ?= -gcflags=all='-N -l'
 GO=GO111MODULE=on GOFLAGS=-mod=vendor go
 GO_BUILD_RECIPE=CGO_ENABLED=0 $(GO) build $(GO_GCFLAGS)
 
+OUT_DIR ?= bin
```
Comment (ironcladlou) [inline Makefile:18]: This change is to make my local dev workflow more flexible, could be extracted into a new PR

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+var (
+	ComponentLabel = "app"
+	NamespaceLabel = "namespace"
```
Comment (ironcladlou) [inline control-plane-operator/controllers/metrics/metricscontroller.go:27]: This may be redundant given the metrics are assigned a `namespace` label during scrape, meaning this value ends up in the `exported_namespace` field as a conflict with a duplicate value

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 5 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig:     &tls.Config{InsecureSkipVerify: true},
+		},
+	)
+	promClient, err := promapi.NewClient(promapi.Config{
```
Comment (ironcladlou) [inline control-plane-operator/controllers/metrics/metricscontroller.go:98]: TODO: recycle client while token is valid

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return err
+	}
+	v1api := promv1.NewAPI(promClient)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	now := time.Now()
+	r := promv1.Range{
+		Start: now.Add(-1 * time.Minute),
```
Comment (ironcladlou) [inline control-plane-operator/controllers/metrics/metricscontroller.go:112]: TODO: figure out the best range to use

File: control-plane-operator/main.go [hunk 6]
```diff
@@ -180,6 +182,17 @@ func NewStartCommand() *cobra.Command {
 			os.Exit(1)
 		}
 
+		podCache, err := cache.New(mgr.GetConfig(), cache.Options{
+			Scheme:    hyperapi.Scheme,
+			Namespace: namespace,
+		})
+		metricsController := metrics.New(podCache, namespace, setupLog)
+		go func() {
+			if err := metricsController.Run(ctx, 30*time.Second); err != nil {
```
Comment (ironcladlou) [inline control-plane-operator/main.go:191]: TODO: Settle on interval

File: hack/metrics/user-workload-monitoring-config.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: user-workload-monitoring-config
+  namespace: openshift-user-workload-monitoring
+data:
+  config.yaml: |
+    prometheus:
+      remoteWrite:
+      - url: https://prometheus-prod-10-prod-us-central-0.grafana.net/api/prom/push
```
Comment (ironcladlou) [inline hack/metrics/user-workload-monitoring-config.yaml:10]: TODO: parameterize this

File: hypershift-operator/controllers/manifests/controlplaneoperator/manifests.go [hunk 8]
```diff
@@ -48,6 +48,14 @@ func OperatorClusterRoleBinding(controlPlaneOperatorNamespace string) *rbacv1.Cl
 	}
 }
 
+func OperatorMonitoringClusterRoleBinding(controlPlaneOperatorNamespace string) *rbacv1.ClusterRoleBinding {
```
Comment (ironcladlou) [inline hypershift-operator/controllers/manifests/controlplaneoperator/manifests.go:51]: TODO: Add a note that this needs replaced with a namespace scoped role binding if possible; according to the monitoring team it might be possible

File: hack/metrics/user-workload-monitoring-config.yaml (hunk 7 above)
Comment (ironcladlou) [reply hack/metrics/user-workload-monitoring-config.yaml:10]: @alvaroaleman what's your feeling on whether this stuff (and/or the scripts) should be ported to Go instead of shell/YAML? The moment I find myself needing to parameterize YAML (and through Bash) I start to think about whether it's time to move to Go, but that could also be a followup refactor... (cc @csrwng)

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+var (
+	ComponentLabel = "app"
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/metrics/metricscontroller.go:26]: Nit: use a `const`

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+	ComponentMemoryRSS = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/metrics/metricscontroller.go:28]: Not a big fan of package-level vars, could we maybe make this part of the controller and register it in the constructor?

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+		return nil, fmt.Errorf("pod cache is required")
+	}
+	if opts.Log == nil {
+		return nil, fmt.Errorf("log is required")
+	}
+	return opts, nil
+}
+
+// Run starts collecting metrics on interval until ctx is closed. It blocks.
+func (c *MetricsController) Run(ctx context.Context, interval time.Duration) error {
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/metrics/metricscontroller.go:98]: Why is the `interval` configured here and not in `NewController`? Removing that as arg would also make this a `Runnable` and allow the `Manager` to run it

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+	log                  logr.Logger
+}
+
+// Options configure the metrics controller.
+type Options struct {
+	// Namespace is the namespace of the hosted control plane.
+	Namespace string
+
+	// PodCache should provide access to pods in Namespace.
+	PodCache cache.Cache
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/metrics/metricscontroller.go:57]: Can we make this a `crclient.Reader` so testing doesn't entail building a fake cache? The cache implements that interface.

File: control-plane-operator/controllers/metrics/metricscontroller.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+		if !hasPodLabel {
+			c.log.V(1).Info("skipping metric missing pod label", "metric", sample.Metric)
+			continue
+		}
+		if len(sample.Values) == 0 {
+			c.log.V(1).Info("skipping metric missing value", "metric", sample.Metric)
+			continue
+		}
+		var pod corev1.Pod
+		if err := c.podCache.Get(ctx, crclient.ObjectKey{Namespace: c.namespace, Name: string(podName)}, &pod); err != nil {
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/metrics/metricscontroller.go:187]: Maybe tolerate 404 here? (Tolerate as in not log an error but still `continue`)

File: hack/metrics/cluster-monitoring-config.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cluster-monitoring-config
+  namespace: openshift-monitoring
+data:
+  config.yaml: |
+    enableUserWorkload: true
```
Comment (alvaroaleman) [inline hack/metrics/cluster-monitoring-config.yaml:8]: Out of curiosity, does the value of `config.yaml` get merged if the cluster already has something? Presumambly not, right?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 15]
```diff
@@ -2703,6 +2735,15 @@ func (r *HostedClusterReconciler) delete(ctx context.Context, req ctrl.Request,
 		}
 	}
 
+	binding := controlplaneoperator.OperatorClusterRoleBinding(controlPlaneNamespace)
+	if err := r.Delete(ctx, binding); err != nil {
```
Comment (alvaroaleman) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2739]: Delete?

**Review by alvaroaleman — COMMENTED** [review]
A couple of comments, let me know what you think. If possible I would really prefer distinct PRs for "Add the controller" and "Configure CI to forward metrics to grafana cloud"

File: control-plane-operator/controllers/hostedcontrolplane/pki/cert.go (hunk 1 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/pki/cert.go:39]: same here, we might be missing something in the createOrUpdate roundtrip. FWIW  SecretTypeOpaque is default, we don't even need to explicitly set it.

**Comment by enxebre** [comment]
@ironcladlou two main questions:
1 - Do we really need a transformation for container_memory_rss -> hypershift_component_memory_rss? Or this something we'd expect a metrics consumer (e.g grafana) to obtain via query.

2 - If we think we do, why implementing a custom controller instead of implement the prometheus collector interface, which seems to fit exactly this use case. https://github.com/prometheus/client_golang/blob/master/prometheus/collector.go#L16-L27

**Comment by ironcladlou** [comment]
@csrwng @enxebre @alvaroaleman 

We might be able to get rid of the controller by implementing a new PrometheusRule like:

```
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: metrics
  namespace: hypershift
spec:
  groups:
  - name: hypershift-control-plane.rules
    interval: 20s
    rules:
    - record: hypershift:controlplane:component_memory_usage
      expr: |
        sum by (label_app, namespace) (
          sum(container_memory_usage_bytes) by (pod, namespace)
        * on (pod) group_left(label_app)
          kube_pod_labels{label_hosted_control_plane!=""}
        )
```

And then simply configuring UWM to export `hypershift:controlplane:component_memory_usage` with any additional transformations. Unfortunately I haven't been able to get the monitoring stack to process my custom rule, which I'm still debugging.

I also discovered a `kube_namespace_labels` key which should eliminate the need to label pods and join on `kube_pod_labels` by labeling the namespace itself, but as far as I can tell `kube_namespace_labels` isn't actually exposing the label values on any namespace, so it could be broken. I'll have to follow up on that, too. Worst case we label all our pods and keep the `kube_pod_labels` join.

**Comment by ironcladlou** [comment]
Turns out that `kube_namespace_labels` is indeed bugged: https://bugzilla.redhat.com/show_bug.cgi?id=2011698. In the meantime we can join on pod labels and refactor that piece away later as a simplification.

File: control-plane-operator/controllers/hostedcontrolplane/pki/cert.go (hunk 1 above)
Comment (ironcladlou) [reply control-plane-operator/controllers/hostedcontrolplane/pki/cert.go:39]: The cluster where it was reproduced is long gone, haven't investigated. In any case, the new code I have here is technically correct because that field is immutable, so not sure why not be defensive.

**Comment by ironcladlou** [comment]
New implementation doesn't rely on the controller at all, check the PR description.

File: cmd/install/assets/hypershift_operator.go [hunk 16]
```diff
@@ -454,35 +456,69 @@ type HyperShiftServiceMonitor struct {
...
+					Interval: "30s",
+					Port:     "metrics",
+				},
+			},
+		},
+	}
 }
+
+type HypershiftRecordingRule struct {
+	Namespace *corev1.Namespace
```
Comment (alvaroaleman) [inline cmd/install/assets/hypershift_operator.go:487]: might be simpler as `Namespace string`

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 17]
```diff
@@ -404,7 +404,13 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// Reconcile the hosted cluster namespace
 	controlPlaneNamespace := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
-	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, controlPlaneNamespace, NoopReconcile)
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, controlPlaneNamespace, func() error {
+		if controlPlaneNamespace.Labels == nil {
+			controlPlaneNamespace.Labels = make(map[string]string)
+		}
+		controlPlaneNamespace.Labels[hyperv1.ControlPlaneComponent] = "namespace"
```
Comment (alvaroaleman) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:411]: err..Why? That this is a namespace and used for a hostedcluster is implicit or not...?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 18]
```diff
@@ -1235,6 +1241,7 @@ func (r *HostedClusterReconciler) reconcileControlPlaneOperator(ctx context.Cont
 		podMonitor.Spec.Selector = *controlPlaneOperatorDeployment.Spec.Selector
 		podMonitor.Spec.PodMetricsEndpoints = []prometheusoperatorv1.PodMetricsEndpoint{{Port: "metrics"}}
 		podMonitor.Spec.NamespaceSelector = prometheusoperatorv1.NamespaceSelector{MatchNames: []string{controlPlaneNamespace.Name}}
+		podMonitor.Spec.JobLabel = "app"
```
Comment (alvaroaleman) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1244]: why? This will be very confusing if we have `app=namespace/podmonitor/$name` rather than `job=namespace/podmonitor/$name` or not?

File: test/setup/main.go [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+	userWorkloadMonitoringNamespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "openshift-user-workload-monitoring",
+		},
+	}
+	err := wait.PollUntil(1*time.Second, func() (done bool, err error) {
+		err = k.Get(ctx, client.ObjectKeyFromObject(userWorkloadMonitoringNamespace), userWorkloadMonitoringNamespace)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Info("waiting for user workload monitoring namespace to exist", "namespace", userWorkloadMonitoringNamespace.Name)
```
Comment (alvaroaleman) [inline test/setup/main.go:181]: Maybe call it by its name (`openshift-user-workload-monitoring`) rather than a description?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 17 above)
Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:411]: Yes, that this is a hypershift control plane namespace. When `kube_namespace_labels` is fixed, we can adjust the recording rule to join on that instead of `kube_pod_labels` (and then we technically wouldn't need this label on our pods at all)

Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: Changed to hypershift.openshift.io/hosted-control-plane for now

**Review by alvaroaleman — APPROVED** [review]
